### PR TITLE
Update Totem T2 mapping, fix bug in VFAT footprint check, and implement suggestions from PR #41777

### PR DESCRIPTION
**Pull request by oljemark**
#### PR description:

For the TOTEM special foreseen on 25.6.2023, the new T2 telescope will be used (and removed after the run) to measure inelastic rate and veto diffractive p+p collisions, when measuring the elastic p+p cross section.
Previously, PR #41472 introduced the 2-channel unpacking for T2 and PR #41777 included some T2 DQM plots and fixes. As part of the approval for PR #41777 , some fixes and syntax safety improvements were asked for, see https://github.com/cms-sw/cmssw/pull/41777#issuecomment-1564189810 . This meant that the tracked T2 parameter useOlderT2TestFile was added to several python config files in EventFilter/CTPPSRawToDigi, and also some other cosmetic changes to those files. 
Last sunday @perrotta wrote a PR (#41784, #41785 ) to implement the asked-for changes, from which I include most changes into this PR.

Last week we also obtained new T2 test files with a new mapping (the final version), and a more diverse set of emulated T2 data frames. (Run 368080 had LE distributed around 5 and TE ~ 12 for all events, Run 368081 same LE value but TE ~9, and Run 368082 same as 368080 but with tracks in half the planes only). Testing the code with these new files exposed several more bugs in the T2 code in the common PPS unpacker.

On receiving more feedback from the detector experts, it turned out that one more member needed to be added to the T2Digi definition in DataFormats/TotemReco , to access header flags. @vavati also suggested that the DQM additions in PR #41777 needed to be expanded by adding the T2 as a DQMCalibrationSource, necessitating changes to RecoPPS/Configuration, also including the XML T2 geometry in a RecoPPS/Local T2 config file since the geometry was used in both the T2Rechit producer and T2 DQM module, but was not found in the database as far as we looked (thanks,  @grzanka !) . The multichannel unpacker from PR #41472 was also turned on by default, needing era modifiers for 2016-2018 PPS data taken without the T2 detector, for which the dummy mapping file mapping_totem_nt2_2021.xml caused an exception since it was copied from 2017 diamond mapping without the changes in T2 xml schema introduced in PR #41472.

At present the T2 CRC field in the VFAT frame in the test files are still just a constant stub, but in the next version of the T2 firmware expected within a few days this will be calculated, and can validate the present calculation in the code. The code is using the same CRC formula as the RP, but due to the inverted order of reading the VFAT frame between T2 & Roman Pots, this needed new helper functions, as did the VFAT signature 'A','C','E' header word intializer checked by the footprint helper.

#### PR validation:

A subset of  runTheMatrix tests relevant for the PPS detectors that might be affected by our code changes was run successfully after the era modifier fix noted above. Those were 136.793 & 1041 & 1042. The T2 unit test under RecoPPS/Local passed, while previously one unit test in  in CalibPPS/TimingCalibration & 25 DQM/Integration were seen to fail when run on lxplus due to the files being on tape, but succeed when cmsbot/Jenkins ran on locally cached copies, see this comment:  https://github.com/cms-sw/cmssw/pull/41785#issuecomment-1570004633

The PR #41785 still passed the tests even so ( https://github.com/cms-sw/cmssw/pull/41785#issuecomment-1566839728 ), although more changes have been added than is included there.

The T2 unit test in RecoPPS/Local was successfully run with the older definition & xml mapping, producing T2Digis, and likewise the totemt2_dqm_test_nino_cfg.py script on last week's new files, produced the expected DQM plots. The  totemt2_dqm_test_common_cfg.py script was run on PPS alignment run data with the other PPS detectors connected but not T2, showing that even though we include T2 in the common CTPPS RawToDigi and Reconstruction tasks, this does not negatively affect the other PPS detectors.
[DQM_V0001_CTPPS_R000366192.zip](https://github.com/cms-sw/cmssw/files/11638296/DQM_V0001_CTPPS_R000366192.zip)

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

These three PR's, #41472 , #41777 and this one should be backported either to a special release on the 13_0_X cycle or the appropriate regular release on the 13_0_X . At a lower priority, backporting will also be done to 13_1_X.

### CODE REVIEW

File: CalibPPS/ESProducers/python/totemT2DAQMapping_cff.py [hunk 1]
```diff
@@ -4,19 +4,25 @@
 
 totemDAQMappingESSourceXML = _xml.clone(
     subSystem = "TotemT2",
-    multipleChannelsPerPayload = cms.untracked.bool(False),
+    multipleChannelsPerPayload = cms.untracked.bool(True),
```
Comment (tvami) [inline CalibPPS/ESProducers/python/totemT2DAQMapping_cff.py:7]: since this is cloned, isnt the `cms.untracked.bool` part unnecessary?

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by grzanka** [comment]
The DQM module for T2 is intended to run in the calibration mode,
to get the plots visible a change is needed here:

https://github.com/cms-sw/cmssw/blob/b2aa6bb85cb8738555dc8ace0100d2c6f1f54b8a/DQM/Integration/python/clients/ctpps_dqm_sourceclient-live_cfg.py#L92

If I understood correctly this change will be done at a later stage by DQM DOC

**Comment by oljemark** [comment]
Hi @tvami . I made the parameter tracked since it could affect the output file (https://github.com/cms-sw/cmssw/pull/41777#issuecomment-1564189810). I also fixed the thrown exception in the unit tests (probably; I didn't find a FEVT dataset that was still on disk to test it locally), but the relVal are still probably failing, since 138.4 did for me.

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by oljemark** [comment]
Hi, @tvami . Some of the relVal errors in 138.4, 138.5, etc seem to be caused by the TotemT2RecHitProducer reading the TotemGeometryRcd here ( https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Local/plugins/TotemT2RecHitProducer.cc#L50 ) from an xml file included here ( https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Local/python/totemT2LocalReconstruction_cff.py ).
This seems to overwrite another record from the database, maybe here ( https://github.com/CTPPS/cmssw/blob/fo_nt2_mapfix2/RecoPPS/Configuration/python/recoCTPPS_cff.py#L13 ) , since the error complains of "Deleting previous geometry: cmsMagneticField:MAGF/Detector Geometry", and 
'Geometry/CMSCommonData/data/muonMagnet.xml' is not included in the T2 xml file.

So how should we include this T2-only TotemGeometryRcd from xml file (to have it in the online DQM calibration chain)? Any advice from the experts?

**Comment by vavati** [comment]
Are we sure that T2 geometry is not already included in the global geometry in DB?

**Comment by bsunanda** [comment]
Totem T2 Geant4 geometry is a part of both 2022(2021) and 2023 geometry. But the reconstruction geometry record is not there, To have that, one needs to add a GeometryWriter in CondTools/Geometry/test - there are writers for all other components. So my request is to make such a writer and test that first before committing to the gitlab.
________________________________

**Comment by oljemark** [comment]
Thanks, @bsunanda ! We're looking into that, to see if & how to make those changes in another PR before the deadline this week for this PR.

**Comment by oljemark** [comment]
Hello, everyone. Since there are some conflicts between the DB-loaded geometry and the XML-loaded T2 geometry (even though no record TotemGeometryRcd is associated with any tag in the DB), @fabferro & @vavati suggested the solving of that should be tested thoroughly so as not to introduce bugs in the PPS geometry, and that within this deadline we should remove the geometry dependence by dropping the T2 DQM dependence on RecHits . So I did that, and also fixed the problem of DQM reading empty T2 Digis, so now both matrix tests 138.4 (delete geometry) & 135.4 (zero products) passed.

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
@cmsbuild , please test

File: CalibPPS/ESProducers/plugins/TotemDAQMappingESSourceXML.cc [hunk 2]
```diff
@@ -276,7 +276,9 @@ TotemDAQMappingESSourceXML::TotemDAQMappingESSourceXML(const edm::ParameterSet &
     : verbosity(conf.getUntrackedParameter<unsigned int>("verbosity", 0)),
       subSystemName(conf.getUntrackedParameter<string>("subSystem")),
       sampicSubDetId(conf.getParameter<unsigned int>("sampicSubDetId")),
-      packedPayload(conf.getUntrackedParameter<bool>("multipleChannelsPerPayload", false)),
+      packedPayload(conf.existsAs<bool>("multipleChannelsPerPayload", true)
+                        ? conf.getParameter<bool>("multipleChannelsPerPayload")
+                        : false),
```
Comment (makortel) [inline CalibPPS/ESProducers/plugins/TotemDAQMappingESSourceXML.cc:281]: The `existsAs()` call could be avoided with implementing `fillDescriptions()`  for this module (https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp)

**Comment by oljemark** [comment]
Hello, @makortel & @fabferro . I added a fillDescription for totemVFATRawToDigi, which passed the RecoPPS/Local T2 unit test, but crashed in PPS matrix test 136.793
[step3_RunDoubleEG2017C.log](https://github.com/cms-sw/cmssw/files/11668569/step3_RunDoubleEG2017C.log)
I'm recompiling with debug symbols to see if I can locate the crash easier. The code is on branch CTPPS:fo_nt2_mapfix2_fillDescPPScrash , the last commit.

**Comment by oljemark** [comment]
Hello again. When I did that & cms-checkdeps, I didn't get a crash anymore. Will add the commit to this branch soon.

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by oljemark** [comment]
Hello @makortel & @tvami . Could you test this 2nd fillDescriptions fix? Looking at all the places the cfi was used, the empty configuration block was always replaced, so this should work (unit test for T2 and PPS matrix tests did)

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by oljemark** [comment]
Hello, everyone. We've been trying to find where the incompatibility between the T2 geometry and some of the relVals, so today we tested adding back the T2 geometry and T2 rechit production into the common CTPPS reconstruction step, see this draft PR: https://github.com/CTPPS/cmssw/pull/76 . This was able to run successfully on PPS alignment data, both in RecoPPS/Local/test/totemT2_geom_rechits_cfg.py (produced PPS classes and empty T2 rechits and digis, since T2 was not included in that run) & DQM/CTPPS/totemt2_dqm_test_common_cfg.py. Testing PPS relVals 136.793 , 1041 & 1042 likewise passed, whereas a recent muon data relval, 140.078 failed in step 3 with the "Deleting previous geometry: cmsMagneticField:MAGF" error seen above. So could the problem be caused by some customisation applied by cmsDriver differing between the Run2_2017 PPS relVal & the Run3 (2022) muon relVal changing the magnetic field or some other property of the forward geometry?

**Comment by oljemark** [comment]
I just tested the relVal 1044 ['RunRawPPS2022B','TIER0EXPPPSCALRUN3','ALCASPLITPPSCALRUN3','ALCAHARVDPPSCALRUN3']  on that same branch, in case moving the PPS to database geometry was the cause of the problem, but that relVal passed without failures.

**Comment by oljemark** [comment]
Here is a newly produced test file with both T2 and TrackingStrip PPS data from that other branch, showing that the reconstruction for both are not mutually incompatible: 
[output-commonReco-368593-1000ev.zip](https://github.com/cms-sw/cmssw/files/11679051/output-commonReco-368593-1000ev.zip)

**Comment by grzanka** [comment]
@makortel I saw interesting thread here: https://github.com/cms-sw/cmssw/issues/33203 which shows similar issue as was discovered with previous commits in this PR, namely: `Deleting previous geometry: cmsMagneticField:MAGF/Detector Geometry`

I suspect the problem stems from the fact that in many PRs we have magnetic field geometry constructed (via `process.load('Configuration.StandardSequences.MagneticField_cff')` injected into some step of relval) On the other hand we've been trying to use the XML source of the T2 detector geometry, via:

```
# T2 geometry seems not to be uploaded to CondDB, here we load XML version, which produces `TotemGeometryRcd`, consumed only by T2 code.
from Geometry.ForwardCommonData.totemT22021V2XML_cfi import *
from Geometry.ForwardGeometry.totemGeometryESModule_cfi import *
```

Do you think it is possible to run the reconstruction (assuming we only need that for special T2 run) on Tier0 where we request the geometry of T2 via XMLIdealGeometryESSource ?

**Comment by tvami** [comment]
> Do you think it is possible to run the reconstruction (assuming we only need that for special T2 run) on Tier0 where we request the geometry of T2 via XMLIdealGeometryESSource ?

This seems pretty complicated since that would mean the introduction of new scenario (T0 does not know about eras, etc, just the scenarios), but not impossible.

**Comment by grzanka** [comment]
> > Do you think it is possible to run the reconstruction (assuming we only need that for special T2 run) on Tier0 where we request the geometry of T2 via XMLIdealGeometryESSource ?
> 
> This seems pretty complicated since that would mean the introduction of new scenario (T0 does not know about eras, etc, just the scenarios), but not impossible.

Thanks ! Can you point to me to anything that can be run locally and is the closest to the T0 execution (some cmsDriver command or similar) ? I would like to check what is possible when running on Tier0 in the scenario when we request the geometry of T2 via XMLIdealGeometryESSource

**Comment by makortel** [comment]
> I saw interesting thread here: #33203 which shows similar issue as was discovered with previous commits in this PR, namely: `Deleting previous geometry: cmsMagneticField:MAGF/Detector Geometry`

My memory is quite rusty, but the issue #33203 was fixed by protecting `DDDetectorESProducer` and `DD4hep_VolumeBasedMagneticFieldESProducerFromDB`, whereas the ESProducers using the `cms::DDCompactView` did not need anything special. If you managed to create a setup where the
```
Deleting previous geometry: cmsMagneticField:MAGF/Detector Geometry
```
exception occurs, that might warrant a specific investigation (separately from this PR).

**Comment by grzanka** [comment]
> 

Thanks @makortel , we encounter such exception when adding on top of this PR following changes:
https://github.com/CTPPS/cmssw/pull/76/files

The exception occurs in many relvals and narrowing it down by any command like: `cmsDriver step3new --conditions auto:run3_data_relval -s RAW2DIGI,L1Reco,RECO --datatier RECO --eventcontent RECO --data --process reRECO --scenario pp --era Run3 -n 100 --filein file:step2.root --fileout file:step3.root`

I've noted that the general `cmsDriver.py` reconstruction was adding magnetic field configuration (`process.load('Configuration.StandardSequences.MagneticField_cff'`) and was curious if that may be somehow related.

On the other hand running PPS reconstruction alone (via `cmsRun`) works well.

I am investigating now how much flexibility we have in running the scenarios for Tier0.

**Comment by grzanka** [comment]
@makortel @tvami while investigation possible options I was trying to limit the RECO step to include only PPS Task. I haven't found good explanation of the `-s` option of `cmsDriver.py` on the TWiki (https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCmsDriver#StandardSteps).

Do you know how whats the syntax to limit the reconstruction to a certain step (in our case PPS reconstruction), is it like `RECO:recoCTPPS` ?

**Comment by tvami** [comment]
> Do you know how whats the syntax to limit the reconstruction to a certain step (in our case PPS reconstruction), is it like RECO:recoCTPPS ?

Yes that should work, I certainly used `RECO:reconstruction_trackingOnly` in the past.

>  I would like to check what is possible when running on Tier0 in the scenario when we request the geometry of T2 via XMLIdealGeometryESSource

The scenarios are defined here:

https://github.com/cms-sw/cmssw/tree/master/Configuration/DataProcessing/python/Impl

**Comment by grzanka** [comment]
@tvami @makortel from our side we are not planning to add any more commit to this PR. Could we proceed with the review now ?
This PR in the current state allows us to get the data from nT2 as DIGI and is not crashing any relvals. 

In parallel we will investigate possible options to include the RecHits for T2, but the most important issue now is to get this PR merged, as without it we won't be able to take any data.

File: DataFormats/TotemReco/interface/TotemT2Digi.h [hunk 3]
```diff
@@ -12,12 +12,13 @@
...
 public:
   TotemT2Digi() = default;
-  TotemT2Digi(unsigned short id, unsigned char marker, unsigned short le, unsigned short te);
+  TotemT2Digi(unsigned short id, unsigned char marker, unsigned short le, unsigned short te, unsigned char status);
 
   void setLeadingEdge(unsigned short le) { lead_edge_ = le; }
   unsigned short leadingEdge() const { return lead_edge_; }
   void setTrailingEdge(unsigned short te) { trail_edge_ = te; }
   unsigned short trailingEdge() const { return trail_edge_; }
+  unsigned char getStatus() const { return status_ & 0xF; }
```
Comment (perrotta) [inline DataFormats/TotemReco/interface/TotemT2Digi.h:21]: This is very minor, but if you need to implement any other modification to the PR, the name of this method should be preferably `status()`, without the "get" (see item 11 in http://cms-sw.github.io/cms_coding_rules.html#2--naming-rules-1)

**Comment by perrotta** [comment]
urgent

**Comment by perrotta** [comment]
@cms-sw/reconstruction-l2 @cms-sw/alca-l2 @cms-sw/dqm-l2 the backport of this PR needs to be included in a special release for the Totem run of June 25. As you may understand, times are pretty short. Could you please try to review this at your earliest, in order to have the 13_0_X backport ready sometime at the beginning of next week, and the release right after (if everything is ok)? Thank you!

**Comment by mandrenguyen** [comment]
+1

**Comment by grzanka** [comment]
> @cms-sw/reconstruction-l2 @cms-sw/alca-l2 @cms-sw/dqm-l2 the backport of this PR needs to be included in a special release for the Totem run of June 25. As you may understand, times are pretty short. Could you please try to review this at your earliest, in order to have the 13_0_X backport ready sometime at the beginning of next week, and the release right after (if everything is ok)? Thank you!

Thank you for prompt action. How should we deal technically with the backport ? Should it go as usual to `13_0_X` ?

**Comment by perrotta** [comment]
> Thank you for prompt action. How should we deal technically with the backport ? Should it go as usual to `13_0_X` ?

Yes, please prepare a backport for the 13_0_X queue as soon as you feel confident enough that this PR will not require any more significant updates

**Comment by grzanka** [comment]
> > Thank you for prompt action. How should we deal technically with the backport ? Should it go as usual to `13_0_X` ?
> 
> Yes, please prepare a backport for the 13_0_X queue as soon as you feel confident enough that this PR will not require any more significant updates

We can prepare now a new branch with squashed commits, ready for backport. Should we wait for any comments from other teams on this PR who needs to sign (AlCa , DQM) before making backport ? If any new comment appear, we could always cherry-pick the changes.

**Comment by tvami** [comment]
I can sign now, although I would have preferred to have the commits  squashed here too

**Comment by emanueleusai** [comment]
+1

**Comment by perrotta** [comment]
@mandrenguyen  since your [last signature](https://github.com/cms-sw/cmssw/pull/41859#issuecomment-1584380252) this PR was just
- renamed one method, as I asked in https://github.com/cms-sw/cmssw/pull/41859#pullrequestreview-1471623173 (two lines changed with a renamed method)
- all commits squashed

I would like this PR being tested in a few master IBs before getting backported in 13_0_X for the Totem run. Therefore, I will consider your previous @cms-sw/reconstruction-l2 signature as good, and merge this now in time for today's 1100 IB.

Please comment if you think there is anything else that need to be considered, instead.

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
merge

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**